src/coursescheduler/models.py

```diff
@@ -1,59 +1,110 @@
-        "pengRequired": "boolean",
-        "yearRequired": "int",
-        "semester": "string",
-        "professor": "professor object",
-        "timeSlots": "list of timeSlot objects",
-        "academicYear": "int"
+    "preferredCoursesPerSemester": {
+        "fall": And(int, lambda n: n >= 0),
+        "spring": And(int, lambda n: n >= 0),
+        "summer": And(int, lambda n: n >= 0),
     },
-}
-
-# A dictionary of professors where the professor ids are the keys and the values are the attributes of each professor
-professors = {
-    "id": {
-        "name": "string",
-        "isPeng": "boolean",
-        "facultyType": "string",
-        "coursePreferences": "list of coursePreference objects",
-        "teachingObligations": "int",
-        "preferredTimes": "dictionary of DayTimes objects",
-        "preferredCoursesPerSemester": "dictionary of fall, spring, and summer ints",
-        "preferredNonTeachingSemester": "string",
-        "preferredCourseDaySpreads": "a course day spread object"
+    "preferredNonTeachingSemester": Or("FALL", "SPRING", "SUMMER", None),
+    "preferredCourseDaySpreads": ["TWF", "MTh", "M", "T", "W", "Th", "F"],
+})
+
+Course = Schema({
+    "code": str,
+    "title": str,
+    "pengRequired": {
+        "fall": bool,
+        "spring": bool,
+        "summer": bool,
     },
-}
-
-# Basic example of the time slot configuration.
-# Below is a possible solution change if the time slot dictionary is too slow
-# Conflict determination: IF list is same length, only need to compare 1 value
-# First check if the day overlaps, then check if time overlaps
-# MR and TWF will never need to be compared
-timeslots = {
-    1: [["TUESDAY", "830", "920"], ["WEDNESDAY", "830", "920"], ["FRIDAY", "830", "920"]],
-    2: [["TUESDAY", "900", "950"], ["WEDNESDAY", "900", "950"], ["FRIDAY", "900", "950"]],
-    3: [["MONDAY", "900", "1020"], ["THURSDAY", "900", "1020"]],
-    4: [["TUESDAY", "900", "1150"]]
-}
-
-
-# conflicts = {
-#     1: ABC,
-#     2: BC
-# }
-# set_1.isdisjoint(set_2)
-# domain = [1, 2, 3, 4, 5, . . . . . ]
-
-# time_conflicts = {(id1, id2) : bool does_conflict}
+    "yearRequired": Or(1, 2, 3, 4),
+})
+
+TimeSlot = Schema({
+    "dayOfWeek": Or("MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY"),
+    "timeRange": TimeRange,
+})
+
+ProfessorSlim = Schema({
+    "id": str,
+    "name": str,
+})
+
+CourseSection = Schema({
+    "professor": Or(None, ProfessorSlim),
+    Optional("maxCapacity"): Or(None, int),
+    "capacity": Or(None, int),
+    "timeSlots": And([TimeSlot], lambda x: len(x) in [0, 1, 2, 3])  # And(list, validate_time_slots),
+})
+
+CourseOffering = Schema({
+    "course": Course,
+    "sections": And([CourseSection], lambda x: len(x) in [0, 1, 2]),  # And(list, validate_course_sections),
+})
+
+Schedule = Schema({
+    "fall": [CourseOffering],
+    "spring": [CourseOffering],
+    "summer": [CourseOffering],
+})
+
+
+def validate_schedule(schedule, print_output=True):
```

Comment:
Maybe rename functions to reflect that they are validating the model structure. Current name might lead to misunderstanding that the functions validate with regards to constraints in them as well.

Reply:
Good point. I'm not sure what would be a better name, though. `validate_schedule_object` would be my next thought but you can make the same argument. Let me think about that!

Reply:
What about `validate_schedule_structure`?